### PR TITLE
[CARBONDATA-2723][DataMap] Fix bugs in recreate datamap on table

### PR DESCRIPTION
**Pull request by xuchuanyin**
While we drop datamap/table, the executor side cache for datamap is
stale. So if we recreate the datamap with different index columns, when
we are doing data loading, the cache should be cleaned, otherwise the
DataMapWriterListener will not take effect for the new datamap.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java [hunk 1]
```diff
@@ -431,6 +431,7 @@ public void clearDataMaps(String tableUniqName) {
         }
       }
     }
+    allDataMaps.remove(tableUniqName);
```
Comment (xuchuanyin) [inline core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java]: FYI: This is the first keypoint

File: processing/src/main/java/org/apache/carbondata/processing/datamap/DataMapWriterListener.java [hunk 2]
```diff
@@ -54,6 +54,9 @@
    */
   public void registerAllWriter(CarbonTable carbonTable, String segmentId,
       String taskNo, SegmentProperties segmentProperties) {
+    // clear cache in executor side
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/datamap/DataMapWriterListener.java]: FYI: This is the second keypoint

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7016/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5793/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5768/

**Comment by xuchuanyin** [comment]
retest sdv please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5785/

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapMeta.java [hunk 3]
```diff
@@ -74,6 +74,6 @@ public String getDataMapName() {
   @Override
   public String toString() {
     return "DataMapMeta{" + "dataMapName='" + dataMapName + '\'' + ", indexedColumns="
-        + indexedColumns + ", optimizedOperation=" + optimizedOperation + '}';
+        + getIndexedColumnNames() + ", optimizedOperation=" + optimizedOperation + '}';
```
[the remark this answers is not in the document]
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/datamap/DataMapMeta.java]: OK, change it to stringbuilder

**Comment by xuchuanyin** [comment]
comments are resolved

**Comment by xuchuanyin** [comment]
retest this please

**Comment by brijoobopanna** [comment]
retest sdv please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5863/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5819/

**Comment by xuchuanyin** [comment]
retest sdv please

**Comment by xuchuanyin** [comment]
The error in SDV is not related to this PR：
```
org.apache.carbondata.cluster.sdv.generated.QueriesIncludeDictionaryTestCase.TABLE_DICTIONARY_INCLUDE_PushUP_044
Results do not match for query:
== Results ==
!== Correct Answer - 1 ==   == Spark Answer - 1 ==
![7.452597748961225E39]     [7.452597748961223E39]
      
      at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:495)
      at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
      at org.scalatest.Assertions$class.fail(Assertions.scala:1328)
      at org.scalatest.FunSuite.fail(FunSuite.scala:1555)
      at org.apache.spark.sql.common.util.QueryTest.checkAnswer(QueryTest.scala:75)
      at org.apache.spark.sql.common.util.QueryTest.writeAndCheckAnswer(QueryTest.scala:113)
```

**Comment by ravipesala** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5829/

**Pull request closed without merging**